README.md
# wphpfpm (PHP FastCGI Manager for windows) #

wphpfpm is my first go-lang project for manage php-cgi on Windows.

Since php-cgi can only serve one client at one time, unless you use apache's mod_fcgid, it's really hard to manage.

So I wrote it for myself, mainly because using caddy to test php can only start one php-cgi process. It is too inhuman, and when I want to change the php settings and want to restart php-cgi, I have to manually kill php-cgi.

For performance, please refer to  [BENCHMARK.md](./BENCHMARK.md)



## Features ##

1. wphpfpm is a standalone service, similar to php-fpm under Linux
2. You can create multiple instances for multiple version php-cgi
3. php-cgi can set the maximum number of process
4. wphpfpm can be a windows service or running on console mode.
5. JSON format configuration file

Please download GO  [GO SDK](https://golang.org/)  (version 1.12+) and execute the following command to get wphpfpm.exe

~~~bash
go build
~~~

Or refer to [caddy-wphpfpm-example.md](./caddy-wphpfpm-example.md) for a simple setup of caddy + wphpfpm.

## Configure file  ##

The following is a json example. The source code [config-sample.json](./config-sample.json) can be download and modify it for your environment.

```json
{
    "LogLevel" : "ERROR",
    "Logger": {
        "FileName": "C:\\wphpfpm\\wphpfpm.log",
        "MaxSize":    10,
        "MaxBackups": 4,
        "MaxAge":     7,
        "Compress":   true,
        "Note": "If you don't need Logger, you can remove the entire Logger section, MaxSize is the unit of MB, MaxAge is the unit of days, this example has 7 days of content per log."
    },
    "Instances" : [
        {
            "Bind" : "127.0.0.1:8000",
            "ExecPath": "C:\\PHP7\\php-cgi.exe",
            "Args" : [],
            "Env": [
                "PHPRC=C:\\PHP7",
                "PHP_FCGI_MAX_REQUESTS=5000" ,
                "PHP_INI_SCAN_DIR=c:\\php7\\conf.d"
            ],
            "MaxProcesses" : 4,
            "MaxRequestsPerProcess": 500
        } ,

        {
            "Bind" : "127.0.0.1:8001",
            "ExecPath": "C:\\PHP5\\php-cgi.exe",
            "Args" : [],
            "Env": [
                "PHPRC=C:\\PHP5",
                "PHP_FCGI_MAX_REQUESTS=5000" ,
                "PHP_INI_SCAN_DIR=c:\\php5\\conf.d"
            ],
            "MaxProcesses" : 2,
            "MaxRequestsPerProcess": 500
        }
    ]
}
```

- LogLevel : According to the level, the default is ERROR
  * PANIC
  * FATAL
  * ERROR
  * WARN
  * INFO
  * DEBUG
  * TRACE
- Logger : You can define the Log output to the file. If you don't need it, you can remove it. The output will be Console (stderr).
- Instances : Define how many kinds of php-cgi to start, this can be used as multiple versions

  - Bind : Define what IP and Port to use for this instance. If multiple versions are required, different Instances must be used with different Ports.

  - ExecPath : php-cgi real  path.

  - Args : You can add parameters for execute php-cgi.exe, note that you can't use  -b  parameters
  - Env : Additional environmental variables
  - MaxProcesses : This directive sets the maximum number of php-cgi processes which can be active at one time.
  - MaxRequestsPerProcess : Each php-cgi  process trip can handle up to several requests. This value must be the same or less than Env's environment variable PHP_FCGI_MAX_REQUESTS.
- Note : This field has no effect, just for comment



## Usage ##

### Run in command line mode (console mode) ###

```
wphpfpm run --conf=config.json
```

### Install as Windows Service ###

```
wphpfpm install --conf=c:\wphpfpm\config.json
```

Note that when install as  a Windows Service, you must use administrator privileges to install.

### Remove wphpfpm service ###

```
wphpfpm uninstall
```



### Start and stop wphpfpm service ###

```
wphpfpm start
wphpfpm stop
```

Alternatively, the service under Windows Control Panel\All Control Panel Items\Administrative Tools\Services can also be started or stopped PHP FastCGI Manager for windows

## Author

- Pigo Chu <pigochu@gmail.com>

- Web Site https://www.pigo.idv.tw

## Resouces ##

- Windows Service Control : https://github.com/chai2010/winsvc
- Command Line parser  : https://gopkg.in/alecthomas/kingpin.v2
- Windows Named Pipe : https://github.com/natefinch/npipe
- Article (https://blog.csdn.net/small_qch/article/details/19562661)
- Log/LogLevel : Logrus (https://github.com/sirupsen/logrus)
- Log Rotate : lumberjack (https://github.com/natefinch/lumberjack)

CHANGELOG.md
# wphpfpm Change log #

### v0.01 2018-9-XX ###

- Implement install as windows service
- implement JSON configuration
- Implement Logger to file
  

logformat_test.go
package main

import (
	"bytes"
	"io/ioutil"
	"strings"
	"testing"
	"unsafe"

	log "github.com/sirupsen/logrus"
)

const (
	testTimeFormat = "2006-01-02 15:04:05 -0700"
)

type BufferFormatter struct {
	timeFormat string
}

func (f *BufferFormatter) Format(entry *log.Entry) ([]byte, error) {
	var b *bytes.Buffer
	if entry.Buffer != nil {
		b = entry.Buffer
	} else {
		b = &bytes.Buffer{}
	}
	b.WriteString(entry.Time.Format(f.timeFormat))
	b.WriteString(" [")
	b.WriteString(entry.Level.String())
	b.WriteString("]: ")
	b.WriteString(entry.Message)
	b.WriteByte('\n')
	return b.Bytes(), nil
}

type StringAddFormatter struct {
	timeFormat string
}

func (f *StringAddFormatter) Format(entry *log.Entry) ([]byte, error) {
	var b string
	b = entry.Time.Format(f.timeFormat) + " [" + entry.Level.String() + "]: " + entry.Message + "\n"
	return []byte(b), nil
}

type StringAddUnsafeFormatter struct {
	timeFormat string
}

func (f *StringAddUnsafeFormatter) Format(entry *log.Entry) ([]byte, error) {
	var s string
	s = entry.Time.Format(f.timeFormat) + " [" + entry.Level.String() + "]: " + entry.Message + "\n"
	return *(*[]byte)(unsafe.Pointer(&s)), nil
}

type StringBuilderFormatter struct {
	timeFormat string
}

func (f *StringBuilderFormatter) Format(entry *log.Entry) ([]byte, error) {
	var builder strings.Builder
	builder.WriteString(entry.Time.Format(f.timeFormat))
	builder.WriteString(" [")
	builder.WriteString(entry.Level.String())
	builder.WriteString("]: ")
	builder.WriteString(entry.Message)
	builder.WriteByte('\n')
	return []byte(builder.String()), nil
}

type StringBuilderUnsafeFormatter struct {
	timeFormat string
}

func (f *StringBuilderUnsafeFormatter) Format(entry *log.Entry) ([]byte, error) {
	var builder strings.Builder

	builder.WriteString(entry.Time.Format(f.timeFormat))
	builder.WriteString(" [")
	builder.WriteString(entry.Level.String())
	builder.WriteString("]: ")
	builder.WriteString(entry.Message)
	builder.WriteByte('\n')
	s := builder.String()
	return *(*[]byte)(unsafe.Pointer(&s)), nil
}

func BenchmarkFormatBuffer(b *testing.B) {
	formatter := &BufferFormatter{timeFormat: "2006-01-02 15:04:05 -0700"}
	log.SetFormatter(formatter)
	log.SetLevel(log.DebugLevel)
	log.SetOutput(ioutil.Discard)
	for i := 0; i < b.N; i++ {
		log.Debugf("%d This is a test message blah blah blah blah blah blah blah blah blah blah ...", i)
	}
}

func BenchmarkFormatStringAdd(b *testing.B) {
	formatter := &StringAddFormatter{timeFormat: "2006-01-02 15:04:05 -0700"}
	log.SetFormatter(formatter)
	log.SetLevel(log.DebugLevel)
	log.SetOutput(ioutil.Discard)
	for i := 0; i < b.N; i++ {
		log.Debugf("%d This is a test message blah blah blah blah blah blah blah blah blah blah ...", i)
	}
}

func BenchmarkFormatStringAddUnsafe(b *testing.B) {
	formatter := &StringAddUnsafeFormatter{timeFormat: "2006-01-02 15:04:05 -0700"}
	log.SetFormatter(formatter)
	log.SetLevel(log.DebugLevel)
	log.SetOutput(ioutil.Discard)
	for i := 0; i < b.N; i++ {
		log.Debugf("%d This is a test message blah blah blah blah blah blah blah blah blah blah ...", i)
	}
}

func BenchmarkFormatBuilder(b *testing.B) {
	formatter := &StringBuilderFormatter{timeFormat: "2006-01-02 15:04:05 -0700"}
	log.SetFormatter(formatter)
	log.SetLevel(log.DebugLevel)
	log.SetOutput(ioutil.Discard)
	for i := 0; i < b.N; i++ {
		log.Debugf("%d This is a test message blah blah blah blah blah blah blah blah blah blah ...", i)
	}
}

func BenchmarkFormatBuilderUnsafe(b *testing.B) {
	formatter := &StringBuilderUnsafeFormatter{timeFormat: "2006-01-02 15:04:05 -0700"}
	log.SetFormatter(formatter)
	log.SetLevel(log.DebugLevel)
	log.SetOutput(ioutil.Discard)
	for i := 0; i < b.N; i++ {
		log.Debugf("%d This is a test message blah blah blah blah blah blah blah blah blah blah ...", i)
	}
}


conf/conf.go
package conf

import (
	"encoding/json"
	"io/ioutil"
	"os"
)

// Conf : JSON root
type Conf struct {
	// Instances 為陣列，包含了多個 ConfInstance
	Instances []Instance
	LogLevel  string  `json:"LogLevel"`
	Logger    *Logger `json:"Logger"`
}

// Instance : JSON Instances
type Instance struct {
	Bind     string   `json:"Bind"`
	ExecPath string   `json:"ExecPath"`
	Args     []string `json:"Args"`
	Env      []string `json:"Env"`
	// MaxRequestsPerProcess 每個php-cgi行程最多能夠處理幾次要求 , Default 500
	MaxRequestsPerProcess int `json:"MaxRequestsPerProcess,500"`
	// MaxProcesses 定義 Instance 啟動 php-cgi 的最大數量，default 4
	MaxProcesses int `json:"MaxProcesses,4"`
	// Note 只是註解，此欄位沒有任何作用
	Note string `json:"-"`
}

// Logger : the same lumberjack.Logger
// see https://github.com/natefinch/lumberjack
type Logger struct {
	Filename   string `json:"Filename"`
	MaxSize    int    `json:"MaxSize,10"`
	MaxAge     int    `json:"MaxAge,7"`
	MaxBackups int    `json:"MaxBackups,4"`
	LocalTime  bool   `json:"LocalTime,true"`
	Compress   bool   `json:"Compress,false"`
}

// LoadFile 讀取 JSON 設定檔，並返回 *Conf
func LoadFile(filePath string) (conf *Conf, err error) {

	jsonFile, err := os.Open(filePath)
	if err != nil {
		return
	}
	defer jsonFile.Close()

	conf = &Conf{LogLevel: "ERROR"}
	byteValue, _ := ioutil.ReadAll(jsonFile)
	err = json.Unmarshal(byteValue, &conf)
	return
}

// FileExist config file is exist 檢查檔案是否存在
func FileExist(filePath string) bool {
	if _, err := os.Stat(filePath); err != nil {
		if os.IsNotExist(err) {
			return false
		}
	}
	return true
}


server/sever.go
package server

import (
	"net"

	log "github.com/sirupsen/logrus"
	"golang.org/x/net/netutil"
)

// Server 定義 Server 的一些參數
type Server struct {
	// 自定義 Tag
	Tag interface{}
	// MaxConnections 定義最大連接數量，必須大於 0 , 否則無上限
	MaxConnections int
	// BindAddress 定義要 listen 的 Address 及 Port , 如 127.0.0.1:8000
	BindAddress string
	listener    net.Listener

	shutdownChan chan bool // 此值如果為 true , 代表 Server 必須停止，所有工作都需要關閉
	shutdown     bool
}

// Conn 是當 Accept 後產生的連線物件
type Conn struct {
	net.Conn // 繼承原本的 net.Conn
	ctx      interface{}
	server   *Server // Server
}

// SetContext 設定任意型態的關聯資源 , 可藉由 Context() 取得
func (c *Conn) SetContext(ctx interface{}) {
	c.ctx = ctx
}

// Context 取得關聯資源 , 可藉由 SetContext() 設定
func (c *Conn) Context() interface{} {
	return c.ctx
}

// Server ...
func (c *Conn) Server() *Server { return c.server }

// Serve ...
func (s *Server) Serve(event Event) error {
	var err error
	s.shutdown = false
	s.listener, err = net.Listen("tcp", s.BindAddress)

	if err != nil {
		return err
	}
	log.Debugf("Server %s starting listener", s.BindAddress)

	s.listener = netutil.LimitListener(s.listener, s.MaxConnections)
	var nextAction Action
	nextAction = None

	if event.OnStartup != nil {
		// 如果有實作 Startup 事件，就呼叫
		nextAction = event.OnStartup(s)
	}

	if &nextAction == nil || nextAction == None {
		return s.loopAccept(event)
	}

	return nil
}

// loopAccept 開始接受外部連線
func (s *Server) loopAccept(event Event) error {

	log.Infof("Server %s starting accept", s.BindAddress)
	s.shutdownChan = make(chan bool, 1)

	for {

		select {
		default:
		case <-s.shutdownChan:
			s.triggerOnShutdown(event)
			return nil
		}

		netconn, err := s.listener.Accept()

		if err == nil {
			conn := &Conn{netconn, nil, s}
			if log.IsLevelEnabled(log.DebugLevel) {
				log.Debugf("Accept %s to %s", conn.RemoteAddr().String(), conn.LocalAddr().String())
			}
			go func(c *Conn) {

				nextAction := s.triggerOnConnect(event, c)
				switch nextAction {
				case Close:
					s.triggerOnDisconnect(event, c)
				case Shutdown:
					s.Shutdown()

				}

			}(conn)

		} else {
			if netconn != nil {
				netconn.Close()
			}

			if s.shutdown == true {
				err = nil
			}

			return err
		}

	}

}

func (s *Server) triggerOnConnect(event Event, c *Conn) Action {
	if log.IsLevelEnabled(log.DebugLevel) {
		log.Debugf("Client connect %s to %s", c.RemoteAddr().String(), c.LocalAddr().String())
	}
	nextAction := Close
	if event.OnConnect != nil {
		nextAction := event.OnConnect(c)
		if &nextAction == nil {
			nextAction = Close
		}
	}
	return nextAction
}

func (s *Server) triggerOnDisconnect(event Event, c *Conn) Action {
	if log.IsLevelEnabled(log.DebugLevel) {
		log.Debugf("Client disconnect %s to %s", c.RemoteAddr().String(), c.LocalAddr().String())
	}
	nextAction := None
	c.Close() // 關閉連線
	if event.OnDisconnect != nil {
		nextAction = event.OnDisconnect(c)
		if &nextAction == nil {
			nextAction = None
		}
	}
	return nextAction
}

func (s *Server) triggerOnShutdown(event Event) {
	if event.OnShutdown != nil {
		event.OnShutdown(s)
	}
}

// Shutdown 停止服務
func (s *Server) Shutdown() {
	s.shutdown = true
	close(s.shutdownChan)
	s.listener.Close()
	log.Debugf("Server %s shutdown", s.BindAddress)
}

// Action 定義 Server 接下來的動作
type Action int

const (
	// None action , next default event will be trigger
	None Action = iota
	// Close client connection
	Close
	// Shutdown server
	Shutdown
)

// Event 啟動 Server 時必須實作以下事件
type Event struct {
	// Startup 代表 Server 啟動事件
	OnStartup func(*Server) (action Action)
	// OnConnect 當有 Client 連線成功時觸發 , 所有邏輯可以寫在這
	OnConnect func(*Conn) (action Action)
	// OnDisconnec Client 斷線時觸發
	OnDisconnect func(*Conn) (action Action)
	// OnShutdown Server 要停止前觸發
	OnShutdown func(*Server)
}


phpfpm/phpfpm.go
package phpfpm

import (
	"container/list"
	"sync"
	"wphpfpm/conf"

	log "github.com/sirupsen/logrus"
)

// Instance : struct
type Instance struct {
	execPath  string
	args      []string
	env       []string
	processes []Process
}

var (
	// conf 是 json 讀進來後產生的設定
	phpfpmConf *conf.Conf
	// idleProcesses php-cgi 如果沒有任何連線處理，都存在這
	idleProcesses []*list.List
	stopManage    = false // 如果調用 Stop() , 這個會是 true , 同時 mon() 也不會繼續監控
	mutex         sync.Mutex
)

// Conf : get Json config
func Conf() *conf.Conf {
	return phpfpmConf
}

// Start php-cgi manager
func Start(conf *conf.Conf) (err error) {
	log.Info("phpfpm starting.")
	phpfpmConf = conf
	instanceLen := len(conf.Instances)
	idleProcesses = make([]*list.List, instanceLen)
	for i := 0; i < instanceLen; i++ {
		idleProcesses[i] = list.New()

		for j := 0; j < conf.Instances[i].MaxProcesses; j++ {
			instance := conf.Instances[i]
			p := newProcess(instance.ExecPath, instance.Args, instance.Env)
			p.instanceIndex = i
			err := p.TryStart()
			if err == nil {
				mutex.Lock()
				p.mapElement = idleProcesses[i].PushBack(p)
				mutex.Unlock()
				go monProcess(p)
			} else {
				Stop()
				return err
			}
		}
	}
	log.Info("phpfpm is in loop.")
	return
}

// monProcess 監控 php-cgi 狀態是否跳出
func monProcess(p *Process) {
	log.Infof("Starting monitor php-cgi(%s)", p.ExecWithPippedName())
	defer log.Infof("Stopped monitor php-cgi(%s)", p.ExecWithPippedName())
	for {
		err := p.cmd.Wait()
		if p.requestCount >= phpfpmConf.Instances[p.instanceIndex].MaxRequestsPerProcess {
			err = p.TryStart()
			if err != nil {
				p.restartChan <- false
			} else {
				p.restartChan <- true
			}

			continue
		}
		if err != nil {
			log.Errorf("php-cgi(%s) exit error, because %s", p.ExecWithPippedName(), err.Error())
		}

		mutex.Lock()

		if stopManage {
			// 執行 phpfpm.Stop() 代表不需要再監控了
			mutex.Unlock()
			return
		}

		idleProcesses[p.instanceIndex].Remove(p.mapElement)
		err = p.TryStart()

		if err != nil {
			// 退出監控
			log.Errorf("php-cgi(%s) restart error, because %s", p.ExecWithPippedName(), err.Error())
			mutex.Unlock()
			return
		}
		// 啟動成功
		p.mapElement = idleProcesses[p.instanceIndex].PushBack(p)
		mutex.Unlock()
		if log.IsLevelEnabled(log.InfoLevel) {
			log.Infof("php-cgi(%s) restart successfully.", p.ExecWithPippedName())
		}
	}
}

// Stop php-cgi manager , 所有的 process kill
func Stop() {
	log.Info("phpfpm stoping.")
	stopManage = true
	mutex.Lock()
	defer mutex.Unlock()

	var next *list.Element

	for _, v := range idleProcesses {
		for e := v.Front(); e != nil; e = next {
			next = e.Next()
			p := e.Value.(*Process)
			p.Kill()
			v.Remove(e)
		}

	}
	log.Info("phpfpm stopped.")
}

// GetIdleProcess : 取得任何一個 Idle 的 Process , 並且移除 Idle 列表
func GetIdleProcess(instanceIndex int) (p *Process) {
	mutex.Lock()
	defer mutex.Unlock()
	e := idleProcesses[instanceIndex].Front()
	if e != nil {
		p = idleProcesses[instanceIndex].Remove(e).(*Process)
		p.mapElement = nil
	}
	return
}

// PutIdleProcess : 設定 Process 為 idle
func PutIdleProcess(p *Process) (err error) {

	mutex.Lock()
	defer mutex.Unlock()
	if p.pipe != nil {
		err = p.pipe.Close()
		p.pipe = nil
	}

	if p.requestCount >= phpfpmConf.Instances[p.instanceIndex].MaxRequestsPerProcess {
		log.Warnf("php-cgi(%s) handled %d requests , need restart.", p.execWithPippedName, p.requestCount)
		p.Kill()
		if true == <-p.restartChan {
			p.mapElement = idleProcesses[p.instanceIndex].PushBack(p)
		} else {
			log.Errorf("php-cgi(%s) restart faild.", p.execWithPippedName)
		}
	} else {
		p.mapElement = idleProcesses[p.instanceIndex].PushBack(p)
		if log.IsLevelEnabled(log.DebugLevel) {
			log.Debugf("php-cgi(%s) is idle , requests count : %d", p.execWithPippedName, p.requestCount)
		}
	}
	return
}


phpfpm/process.go
package phpfpm

import (
	"container/list"
	"io"
	"net"
	"os"
	"os/exec"
	"strconv"
	"sync"
	"time"

	log "github.com/sirupsen/logrus"
	"gopkg.in/natefinch/npipe.v2"
)

// Process : struct
type Process struct {
	execPath      string
	args          []string
	env           []string
	cmd           *exec.Cmd
	instanceIndex int // 這個是在 phpfpm.go 中的 idleprocess  連結用的 , 代表這個 Process 是屬於那個 Instance
	mapElement    *list.Element
	pipe          *npipe.PipeConn
	pippedName    string // php-cgi 執行時指定的 pipped name

	requestCount int // 紀錄當前執行中的 php-cgi 已經接受幾次要求了

	restartChan chan bool

	copyRbuf           []byte
	copyWbuf           []byte
	execWithPippedName string

	wg sync.WaitGroup
}

var (
	namedPipeNumber      = time.Now().Unix()
	namedPipeNumberMutex sync.Mutex
)

// newProcess : Create new Process
// 建立一個新的 Process
func newProcess(execPath string, args []string, env []string) *Process {
	p := new(Process)
	p.execPath = execPath
	p.args = args
	p.env = env
	p.restartChan = make(chan bool)
	p.copyRbuf = make([]byte, 4096)
	p.copyWbuf = make([]byte, 16384)
	return p
}

// TryStart will execute php-cgi twince
func (p *Process) TryStart() (err error) {
	// pippedName 是啟動 php-cgi 時候指定 -b pipename 使用的
	namedPipeNumberMutex.Lock()
	namedPipeNumber++
	namedPipeNumberMutex.Unlock()
	p.pippedName = `\\.\pipe\wphpfpm\wphpfpm.` + strconv.FormatInt(namedPipeNumber, 10)
	p.requestCount = 0
	p.execWithPippedName = p.execPath + " -> " + p.pippedName

	log.Debugf("Trying to start php-cgi(%s).", p.execWithPippedName)
	for i := 0; i < 2; i++ {
		args := append(p.args, "-b", p.pippedName)
		p.cmd = nil
		p.cmd = exec.Command(p.execPath, args...)
		p.cmd.Env = os.Environ()
		p.cmd.Env = append(p.cmd.Env, p.env...)
		err = p.cmd.Start()
		if err == nil {
			i = 3
			if log.IsLevelEnabled(log.DebugLevel) {
				log.Debugf("php-cgi(%s) executing now.", p.execWithPippedName)
			}
		}
	}

	if err != nil {
		log.Errorf("php-cgi(%s) can not start , because %s", p.execWithPippedName, err.Error())
	}
	return
}

// connectPipe will connect to php-cgi named pipe
func (p *Process) connectPipe() error {
	var err error
	//if p.pipe != nil {
	//		p.pipe.Close()
	//}

	p.pipe, err = npipe.Dial(p.pippedName)
	if err != nil {
		log.Errorf("Connect to php-cgi(%s) error , because %s", p.execWithPippedName, err.Error())
		return err
	}
	p.requestCount++
	if log.IsLevelEnabled(log.DebugLevel) {
		log.Debugf("Connect to php-cgi(%s) successfully.", p.execWithPippedName)
	}

	return nil
}

// Proxy net.Conn <> Windows-named-pipe
// Proxy 將 tcp 來源跟 windows named pipe 直接做讀寫
// 返回值 serr 代表由 http server 讀取資料寫至 php-cgi 的錯誤
// 返回值 terr 代表由 php-cgi 讀取資料寫至 http server 的錯誤
func (p *Process) Proxy(conn net.Conn) (serr error, terr error) {

	terr = p.connectPipe()
	if terr != nil {
		return
	}

	p.wg.Add(2)
	go func() {
		// read from web server , write to php-cgi
		_, serr = io.CopyBuffer(p.pipe, conn, p.copyRbuf)
		p.wg.Done()
	}()
	go func() {
		// read from php-cgi , write to web server
		_, terr = io.CopyBuffer(conn, p.pipe, p.copyWbuf)
		p.wg.Done()
	}()

	p.wg.Wait()
	p.pipe.Close()
	p.pipe = nil
	return
}

// Kill php-cgi process
func (p *Process) Kill() (err error) {
	err = p.cmd.Process.Kill()

	if err != nil {
		if log.IsLevelEnabled(log.ErrorLevel) {
			log.Errorf("Kill php-cgi(%s) error , because %s", p.execWithPippedName, err.Error())
		}
	} else {
		if log.IsLevelEnabled(log.DebugLevel) {
			log.Debugf("Kill php-cgi(%s) successfully.", p.execWithPippedName)
		}
	}

	return
}

// ExecWithPippedName ...
func (p *Process) ExecWithPippedName() string {
	return p.execWithPippedName
}


main.go
package main

import (
	"bytes"
	"fmt"
	"os"
	"os/signal"
	"path/filepath"
	"strings"
	"sync"
	"wphpfpm/conf"
	"wphpfpm/phpfpm"
	"wphpfpm/server"

	"github.com/chai2010/winsvc"
	log "github.com/sirupsen/logrus"
	"gopkg.in/alecthomas/kingpin.v2"
	"gopkg.in/natefinch/lumberjack.v2"
)

var (
	serviceName      = "wphpfpm"
	serviceDesc      = "PHP FastCGI Manager for windows"
	app              = kingpin.New(serviceName, serviceDesc)
	commandInstall   *kingpin.CmdClause
	commandUninstall *kingpin.CmdClause
	commandStart     *kingpin.CmdClause
	commandStop      *kingpin.CmdClause
	commandRun       *kingpin.CmdClause
	flagConfigFile   *string

	servers []*server.Server
)

func main() {
	if !winsvc.IsAnInteractiveSession() {
		// run as service
		flag := kingpin.Flag("conf", "Config file path , required by install or run.")
		flagConfigFile = flag.Required().String()
		kingpin.Parse()
		checkConfigFileExist(*flagConfigFile)

		fmt.Println(serviceName, "Run service")
		if err := winsvc.RunAsService(serviceName, startService, stopService, false); err != nil {
			log.Fatalf(serviceName+" run: %v\n", err)
		}
	} else {
		// command line mode
		initCommandFlag()
		switch kingpin.Parse() {
		case commandInstall.FullCommand():
			checkConfigFileExist(*flagConfigFile)
			installService()
		case commandUninstall.FullCommand():
			if err := winsvc.RemoveService(serviceName); err != nil {
				fmt.Println("Uninstall service: ", err)
				os.Exit(1)
			}
			fmt.Println("Uninstall service: success")
		case commandRun.FullCommand():
			checkConfigFileExist(*flagConfigFile)
			startService()
		case commandStart.FullCommand():
			if err := winsvc.StartService(serviceName); err != nil {
				fmt.Println("Start service:", err)
				os.Exit(1)
			}
			fmt.Println("Start service: success")
		case commandStop.FullCommand():
			if err := winsvc.StopService(serviceName); err != nil {
				fmt.Println("Stop service:", err)
				os.Exit(1)
			}
			fmt.Println("Stop service: success")
			return
		}
	}

}

func initCommandFlag() {
	commandInstall = kingpin.Command("install", "Install as service")
	commandUninstall = kingpin.Command("uninstall", "Uninstall service")
	commandStart = kingpin.Command("start", "Start service.")
	commandStop = kingpin.Command("stop", "Stop service.")
	commandRun = kingpin.Command("run", "Run in console mode")
	flag := kingpin.Flag("conf", "Config file path , required by install or run.")
	if len(os.Args) > 1 && (os.Args[1] == "install" || os.Args[1] == "run") {
		flagConfigFile = flag.Required().String()
	} else {
		flagConfigFile = flag.String()
	}
}

// 安裝服務
func installService() {
	var serviceExec string
	var err error

	if serviceExec, err = winsvc.GetAppPath(); err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	if err := os.Chdir(filepath.Dir(serviceExec)); err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	abs, err := filepath.Abs(*flagConfigFile)

	serviceExecFull := "\"" + serviceExec + "\"" + " --conf=" + abs
	args := []string{"--conf", abs}
	fmt.Printf("Service install name : %s , binpath : %s\n", serviceName, serviceExecFull)
	if err := winsvc.InstallService(serviceExec, serviceName, serviceDesc, args...); err != nil {
		fmt.Printf("Install service error : %s\n", err.Error())
		os.Exit(1)
	}

	fmt.Println("Install service successfully.")
}

// 啟動服務
func startService() {
	config, err := conf.LoadFile(*flagConfigFile)

	if err != nil {
		fmt.Printf("Config load error : %s\n", err.Error())
		os.Exit(1)
	}

	fmt.Printf("Start in console mode , press CTRL+C to exit ...\r\n")
	initLogger(config)
	err = phpfpm.Start(config)
	if err != nil {
		log.Fatalf("Can not start service : %s\n", err.Error())
	}

	var events server.Event

	events.OnConnect = func(c *server.Conn) (action server.Action) {

		p := phpfpm.GetIdleProcess(c.Server().Tag.(int))

		if p == nil {
			if log.IsLevelEnabled(log.ErrorLevel) {
				log.Error("Can not get php-cgi process")
			}
			action = server.Close
			return
		}
		serr, terr := p.Proxy(c) // blocked
		if log.IsLevelEnabled(log.DebugLevel) {
			log.Debugf("php-cgi(%s) proxy error , serr : %s , terr : %s", p.ExecWithPippedName(), serr, terr)
		}

		phpfpm.PutIdleProcess(p)

		return
	}

	conf := phpfpm.Conf()

	var wg sync.WaitGroup

	wg.Add(len(conf.Instances))

	servers = make([]*server.Server, len(conf.Instances))

	for i := 0; i < len(conf.Instances); i++ {
		instance := conf.Instances[i]
		servers[i] = &server.Server{MaxConnections: instance.MaxProcesses, BindAddress: instance.Bind, Tag: i}

		log.Infof("Start server #%d on %s", i, servers[i].BindAddress)

		go func(s *server.Server) {
			err := s.Serve(events)
			if err != nil {
				log.Errorf("Service serve error : %s", err.Error())
			}
			wg.Done()
		}(servers[i])
	}
	log.Info("Service running ...")

	// 這段處理 CTRL + C
	c := make(chan os.Signal, 1)
	signal.Notify(c, os.Interrupt)
	go func() {
		for sig := range c {
			log.Infof("Service got signal: %s", sig.String())
			stopService()
			return
		}
	}()

	wg.Wait()
	phpfpm.Stop()
	log.Info("Service Stopped.")
}

// 停止服務
func stopService() {

	for i := 0; i < len(servers); i++ {
		servers[i].Shutdown()
	}

}

func checkConfigFileExist(filepath string) {
	exist := conf.FileExist(filepath)
	if !exist {
		fmt.Printf("Could not load config file : %s", filepath)
		os.Exit(1)
	}
}

func initLogger(config *conf.Conf) {

	formatter := &MyTextFormatter{timeFormat: "2006-01-02 15:04:05 -0700"}
	log.SetFormatter(formatter)

	// Set logger
	if len(config.Logger.Filename) > 0 {

		logDir := filepath.Dir(config.Logger.Filename)
		exeDir, err := filepath.Abs(filepath.Dir(os.Args[0])) // 執行檔的路徑
		if err != nil {
			log.Fatal(err)
		}

		if logDir == "." || logDir == "" {
			// 如果 Filename 沒指定路徑，修正為 exe 的路徑
			config.Logger.Filename = exeDir + "\\" + config.Logger.Filename
		}

		logger := &lumberjack.Logger{
			Filename:   config.Logger.Filename,
			MaxSize:    config.Logger.MaxSize,
			MaxBackups: config.Logger.MaxBackups,
			MaxAge:     config.Logger.MaxAge,
			Compress:   config.Logger.Compress,
		}
		log.SetOutput(logger)
		fmt.Printf("Logger ouput set to file %s .\n", config.Logger.Filename)
	} else {
		fmt.Printf("Logger ouput set to console.\n")
	}

	if config.LogLevel == "" {
		config.LogLevel = "ERROR"
	}

	logLevel, err := log.ParseLevel(config.LogLevel)
	if err != nil {
		log.Fatalf("LogLevel %s can not parse.", config.LogLevel)
	}
	log.SetLevel(logLevel)
	log.Infof("Set LogLevel to %s.", strings.ToUpper(logLevel.String()))

	// Repair config
	for i := 0; i < len(config.Instances); i++ {
		if config.Instances[i].MaxRequestsPerProcess < 1 {
			// Repair MaxRequestsPerProcess
			log.Warnf("Instance #%d MaxRequestsPerProcess is less 1 , set to 500", i)
			config.Instances[i].MaxRequestsPerProcess = 500
		}

		if config.Instances[i].MaxProcesses < 1 {
			//Repair MaxProcesses
			log.Warnf("Instance #%d MaxProcesses is less 1 , set to 4", i)
			config.Instances[i].MaxProcesses = 4
		}
	}
}

// MyTextFormatter logrus custom formatter
type MyTextFormatter struct {
	timeFormat string
}

// Format logrus custom format
func (f *MyTextFormatter) Format(entry *log.Entry) ([]byte, error) {
	var b *bytes.Buffer
	if entry.Buffer != nil {
		b = entry.Buffer
	} else {
		b = &bytes.Buffer{}
	}
	b.WriteString(entry.Time.Format(f.timeFormat))
	b.WriteString(" [")
	b.WriteString(entry.Level.String())
	b.WriteString("]: ")
	b.WriteString(entry.Message)
	b.WriteByte('\n')
	return b.Bytes(), nil
}
